### PR TITLE
Add OpenAI Summarize RLHF with trlX

### PR DESCRIPTION
**Pull request by PhungVanDuy**

### CODE REVIEW

File: examples/summarize_rlhf/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+1. Train SFT:
+```bash
+    cd  sft/ && deepspeed train_gptj_summarize.py
+```
+
+Checkpoint: [SFT](https://huggingface.co/pvduy/openai_summarize_sft_gptj)
+
+2. Train Reward Model
+```bash
+    cd reward_model/ && deepspeed train_reward_gptj.py
```
Comment (jon-tow) [inline examples/summarize_rlhf/README.md]: Fix file path name: `train_reward_gptj.py` -> `train_reward_model_gptj.py`

**Review by jon-tow — COMMENTED** [review]
Just leaving some small comments along the way in preparation for a final review. (Looks great though!) I've pushed some very small changes to speed things up @PhungVanDuy

File: examples/summarize_rlhf/trlx_inference_gptj.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    count = 0
+    for post, summarize in tqdm(
+        zip(test_post_list, test_summ_list), total=len(test_post_list)
+    ):
+        encode_dict = tokenizer(
+            post, return_tensors="pt", padding=False, truncation=True
+        )
+        txt_tokens = encode_dict["input_ids"].cuda()
+        attention_mask = encode_dict["attention_mask"].cuda()
+        kwargs = {"max_new_tokens": 50, "eos_token_id": 50256, "pad_token_id": 50256}
```
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply examples/summarize_rlhf/trlx_inference_gptj.py:73]: The paper report T=0, which means greedy sampling iirc and I also tried a few experiments with other sampling methods with top-K and top-P it did not work well like this.

File: examples/summarize_rlhf/sft/train_gptj_summarize.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    #    tokenizer.padding_side = "left"
+    #    tokenizer.truncation_side = "left"
+    model.resize_token_embeddings(len(tokenizer))
+    tokenizer.pad_token_id = tokenizer.eos_token_id
+    model.config.end_token_id = tokenizer.eos_token_id
+    model.config.pad_token_id = model.config.eos_token_id
+    train_dataset = TLDRDataset(
+        "pvduy/openai_summarize_tldr",
+        tokenizer,
+        "train_sft",
```
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply examples/summarize_rlhf/sft/train_gptj_summarize.py]: yes I changed it.

File: examples/summarize_rlhf/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+| Model | Average Reward | Reward $\Delta$ |
+| --- | --- | --- |
+| SFT | 2.457 | -0.406 |
+| PPO | 3.248 | +0.385 |
+
+
+2. Reward model accuracy:
+
+![image](https://user-images.githubusercontent.com/28798474/210157656-c5b20b9a-f6ef-4e88-a0ee-5596d5b28d58.png)
```
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply examples/summarize_rlhf/README.md:49]: We can remove this part, this plot I made just to check how accuracy because when I am training I did not use a full validation set (time inference).

File: examples/summarize_rlhf/trlx_inference_gptj.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply examples/summarize_rlhf/trlx_inference_gptj.py:73]: In our final human evaluations, we use T=0 to sample from all models, as we found it performed better than higher temperatures or nucleus sampling (see Appendix B.1). - Page 5 from the paper.

File: trlx/orchestrator/ppo_orchestrator.py [hunk 5]
```diff
@@ -65,6 +65,7 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
         stats = {}
         clock = Clock()
         while len(ppo_rl_elements) < num_rollouts:
+            print(f"Making experience {len(ppo_rl_elements)} / {num_rollouts}")
```
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply trlx/orchestrator/ppo_orchestrator.py]: will remove it just for debugging purpose forgot to remove :)

File: examples/summarize_rlhf/README.md [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+3. Examples of generated summaries: [here](https://wandb.ai/pvduy/trlx/runs/1rpm40g8)
```
Comment (jon-tow) [inline examples/summarize_rlhf/README.md:51]: This link gives a 404. Maybe we can just sample a few summaries and paste them here or wait until the blog post is ready to link to?

Comment (PhungVanDuy) [reply examples/summarize_rlhf/README.md:51]: @jon-tow https://wandb.ai/carperai/summarize_RLHF/runs/2uirt89a, new one!

**Comment by jon-tow** [comment]
My recent commit cleans up some nits:
* Fixes function name misspelling `create_comparision_dataset` -> `create_comparison_dataset`
* Renames some variable names to be more descriptive (e.g. `lst_post` -> `posts`)


Final TODOs before merging:

- [ ] Replace HuggingFace data path names with CarperAI ones (e.g. `pvduy/openai_summarize_sft_gptj` -> `CarperAI/openai_summarize_sft_gptj`) if we want to use the CarperAI account.

- [ ] Add the blog post link to README.md as it covers the example in detail.

File: trlx/orchestrator/ppo_orchestrator.py [hunk 7]
```diff
@@ -206,21 +207,23 @@ def make_experience(self, num_rollouts: int = 1024, iter_count: int = 0):  # noq
...
-                )  # left shift by 1 ref: https://github.com/lvwerra/trl/blob/main/trl/trainer/ppo_trainer.py#L425
-                ends = start + attention_mask[:, start:].sum(1) - 1
-                for ix in range(n):
-                    if ends[ix] == all_tokens.shape[1]:
-                        ends[ix] = ends[ix] - 1
-                all_values = [values[ix, start - 1 : ends[ix] - 1] for ix in range(n)]
+
+                start = query_tensors.shape[1] - 1
+                ends = start + attention_mask[:, start:].sum(1)
+                all_values = [values[ix, start : ends[ix]] for ix in range(n)]
```
Comment (jon-tow) [inline trlx/orchestrator/ppo_orchestrator.py:224]: We probably should leave this out until we're 100% sure this is the right fix - we need unit tests for these.

**Review by jon-tow — CHANGES_REQUESTED** [review]

File: trlx/orchestrator/ppo_orchestrator.py (hunk 7 above)
Comment (PhungVanDuy) [reply trlx/orchestrator/ppo_orchestrator.py:224]: I do agree but we should have work version at least for simple tasks.

File: trlx/utils/modeling.py [hunk 8]
```diff
@@ -489,3 +489,86 @@ def range_to_pattern(start, stop):
         negative_only_subpatterns + intersected_subpatterns + positive_only_subpatterns
     )
     return "|".join(subpatterns)
+
+
+def best_of_n_sampling(
```
Comment (jon-tow) [inline trlx/utils/modeling.py:494]: Why are we adding `best_of_n_sampling` here if unused in the OpenAI Summarize,  example?  Do we want to squeeze a new feature in here? CC @LouisCastricato

File: trlx/utils/modeling.py [hunk 9]
```diff
@@ -489,3 +489,86 @@ def range_to_pattern(start, stop):
...
+    model,
+    tokenizer,
```
Comment (jon-tow) [inline trlx/utils/modeling.py:496]: Remove, unused arg.

**Review by jon-tow — CHANGES_REQUESTED** [review]

File: trlx/orchestrator/ppo_orchestrator.py (hunk 7 above)
Comment (jon-tow) [reply trlx/orchestrator/ppo_orchestrator.py:224]: Yeah, that's true. If this is just a revert of the stuff introduced in the T5 then we should keep it other I'll defer @LouisCastricato for second opinions.

Comment (LouisCastricato) [reply trlx/orchestrator/ppo_orchestrator.py:224]: Leave for now, fix later

File: trlx/utils/modeling.py (hunk 8 above)
Comment (LouisCastricato) [reply trlx/utils/modeling.py:494]: I wanted to add it in @reciprocated's generation utils pr but it didn't end up happening

Comment (reciprocated) [reply trlx/utils/modeling.py:494]: It was added here on tuesday and I thought Duy had planned to make use of it. I can host it on my pr instead

File: examples/summarize_rlhf/reward_model/reward_model.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                continue
+
+            # Check if there is any padding otherwise take length of sequence
+            c_inds = (chosen[i] == self.PAD_ID).nonzero()
+            c_ind = c_inds[0].item() if len(c_inds) > 0 else chosen.shape[1]
+            r_inds = (rejected[i] == self.PAD_ID).nonzero()
+            r_ind = r_inds[0].item() if len(r_inds) > 0 else rejected.shape[1]
+            end_ind = max(c_ind, r_ind)
+
+            # Retrieve first index where trajectories diverge
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/reward_model/reward_model.py:81]: Maybe we should explain what this means in more detail.

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+import torch
+from datasets import load_dataset
+from reward_model import GPTRewardModel
+from torch.utils.data import Dataset
+from tqdm import tqdm
+from transformers import AutoTokenizer, Trainer, TrainingArguments
+
+
+def create_comparison_dataset(
+    path="pvduy/openai_summarize_comparisions", split="train"
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/reward_model/train_reward_model_gptj.py]: Fix path.

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    dataset = load_dataset(path, split=split)
+    pairs = []
+    for sample in tqdm(dataset):
+        pair = {}
+        prompt = sample["prompt"]
+        chosen_summary = sample["chosen"]
+        rejected_summary = sample["rejected"]
+        if chosen_summary == rejected_summary:
+            continue
+        if len(chosen_summary.split()) < 5 or len(rejected_summary.split()) < 5:
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/reward_model/train_reward_model_gptj.py:23]: Need some comments here.

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    return pairs
+
+
+class PairwiseDataset(Dataset):
+    def __init__(self, pairs, tokenizer, max_length):
+        self.chosen_input_ids = []
+        self.chosen_attn_masks = []
+        self.rejected_input_ids = []
+        self.rejected_attn_masks = []
+        for pair in tqdm(pairs):
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/reward_model/train_reward_model_gptj.py:37]: Likewise, need some comments here too.

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+
+if __name__ == "__main__":
+    tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")
+    tokenizer.pad_token = tokenizer.eos_token
+
+    if not os.path.exists("rm_checkpoint"):
+        os.mkdir("rm_checkpoint")
+
+    training_args = TrainingArguments(
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/reward_model/train_reward_model_gptj.py:97]: Can you not just use a config file for this?

File: examples/summarize_rlhf/sft/ds_config_gptj.json [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+    },
+    "allgather_partitions": true,
+    "allgather_bucket_size": 5e8,
+    "contiguous_gradients": true
+  },
+  "optimizer": {
+    "type": "AdamW",
+    "params": {
+      "lr": 1e-05,
+      "betas": [
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/sft/ds_config_gptj.json:24]: These are the betas i would have expected for the reward model.

File: examples/summarize_rlhf/sft/summarize_dataset.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+        self.post_list = []
+        self.summaries_0 = []
+        self.summaries_1 = []
+        self.labels = []
+        self.max_length = max_length
+
+        def make_text(post, summarize):
+            return f"SUBREDDIT: r/{post['subreddit']}\nTITLE: {post['title']}\nPOST: {post['post']}\nTL;DR: {summarize}"
+
+        for sample in dataset:  # chosen summary is always the first one
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/sft/summarize_dataset.py:74]: Do not put a comment on the same line as the for loop like this. Also, some more explanation here would be nice. This comment does not make much sense.

File: examples/summarize_rlhf/sft/train_gptj_summarize.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    Trainer,
+    TrainingArguments,
+    default_data_collator,
+)
+
+try:
+    import evaluate
+except ImportError:
+    raise ImportError(
+        "To run this example, please install the following packages: [`evaluate`, `nltk`, `rouge_score`]."
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/sft/train_gptj_summarize.py:18]: We should have a requirements.txt for this example, no?

File: examples/summarize_rlhf/sft/train_gptj_summarize.py [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+def set_seed(seed_val=42):
+    random.seed(seed_val)
+    np.random.seed(seed_val)
+    torch.manual_seed(seed_val)
+    torch.cuda.manual_seed_all(seed_val)
+
+
+if __name__ == "__main__":
+    output_dir = "gptj-supervised-summarize-checkpoint"
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/sft/train_gptj_summarize.py:30]: Config file pls

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+from trlx.data.configs import TRLConfig
+
+REWARD_CHECKPOINT_PATH = "reward_model/rm_checkpoint/pytorch_model.bin"
+if not os.path.exists(REWARD_CHECKPOINT_PATH):
+    os.makedirs("reward_model/rm_checkpoint", exist_ok=True)
+    os.system(
+        f"wget -O {REWARD_CHECKPOINT_PATH} \
+        https://huggingface.co/pvduy/openai_summarize_rm_checkpoint/resolve/main/pytorch_model.bin"
+    )
+SFT_MODEL_PATH = "pvduy/openai_summarize_sft_gptj_full_data"
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/trlx_gptj_text_summarization.py]: Fix path so it isn't using your name.

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+    rw_tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")
+    rw_tokenizer.pad_token = rw_tokenizer.eos_token
+    rw_model = GPTRewardModel(SFT_MODEL_PATH)
+    rw_model.load_state_dict(torch.load(REWARD_CHECKPOINT_PATH))
+    rw_model.half()
+    rw_model.eval()
+    rw_device = torch.device("cuda:{}".format(1))
+    rw_model.to(rw_device)
+
+    def reward_fn(samples: List[str]):
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/trlx_gptj_text_summarization.py:35]: We need way more comments here. This is extremely confusing.

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            input_ids = input_ids.repeat(2, 1)
+            attn_masks = attn_masks.repeat(2, 1)
+            with torch.no_grad():
+                sub_scores = rw_model(input_ids=input_ids, attention_mask=attn_masks)
+            original_scores_list.append(sub_scores["chosen_end_scores"])
+        original_scores = torch.cat(original_scores_list, dim=0)
+
+        scores_list = []
+        batch_size = 2
+        for i in range(0, len(samples), batch_size):
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/trlx_gptj_text_summarization.py:66]: Shouldn't we be abstracting away these loops a bit? They look like repeats.

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+    train_set = [(sample["prompt"], sample["label"]) for sample in dataset["train"]]
+    val_set = [(sample["prompt"], sample["label"]) for sample in dataset["valid"]]
+
+    # Split contents into summaries and labels
+    train_openai_summ, train_labels = zip(*train_set)
+    val_openai_summ, val_labels = zip(*val_set)
+
+    train_post_summ = {}
+    train_prompts = []
+    for i in tqdm(range(len(train_openai_summ))):
```
Comment (LouisCastricato) [inline examples/summarize_rlhf/trlx_gptj_text_summarization.py:108]: This should be a helper function in a utility file, not in the train file.

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py (hunk 19 above)
Comment (jon-tow) [reply examples/summarize_rlhf/trlx_gptj_text_summarization.py]: Done in most recent commit

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py (hunk 11 above)
Comment (jon-tow) [reply examples/summarize_rlhf/reward_model/train_reward_model_gptj.py]: Done in most recent commit

File: examples/summarize_rlhf/sft/train_gptj_summarize.py (hunk 18 above)
Comment (jon-tow) [reply examples/summarize_rlhf/sft/train_gptj_summarize.py:30]: Note that this is the `SFT` training script that is `transformers.Trainer`-based. I don't believe `transformers` supplies a config-to-[`transformers.TrainingArguments`](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments) converter, so I'm okay leaving this for now.

File: examples/summarize_rlhf/trlx_gptj_text_summarization.py (hunk 19 above)
Comment (jon-tow) [reply examples/summarize_rlhf/trlx_gptj_text_summarization.py]: Fixed in recent commit

File: examples/summarize_rlhf/reward_model/train_reward_model_gptj.py (hunk 14 above)
Comment (jon-tow) [reply examples/summarize_rlhf/reward_model/train_reward_model_gptj.py:97]: > Note that this is the SFT training script that is transformers.Trainer-based. I don't believe transformers supplies a config-to-[transformers.TrainingArguments](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments) converter, so I'm okay leaving this for now.

See the related comment in the conversation below.

**Review by jon-tow — APPROVED** [review]
